src/DevTools/BlockSources.php

```diff
@@ -0,0 +1,184 @@
+		if ( version_compare( get_bloginfo( 'version' ), '5.5', '<' ) ) {
+			return false;
+		}
+
+		return is_admin() || wp_doing_ajax() || wp_doing_cron() || ( defined( 'REST_REQUEST' ) && REST_REQUEST );
+	}
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param PluginRegistry        $plugin_registry Plugin registry instance.
+	 * @param LikelyCulpritDetector $likely_culprit_detector Likely culprit detector instance.
+	 */
+	public function __construct( PluginRegistry $plugin_registry, LikelyCulpritDetector $likely_culprit_detector ) {
+		$this->plugin_registry         = $plugin_registry;
+		$this->likely_culprit_detector = $likely_culprit_detector;
+	}
+
+	/**
+	 * Runs on instantiation.
+	 */
+	public function register() {
+		$this->set_block_sources_from_cache();
+
+		if ( empty( $this->get_block_sources() ) ) {
+			add_filter( 'register_block_type_args', [ $this, 'capture_block_type_source' ] );
+
+			// All blocks should be registered well before admin_enqueue_scripts.
+			add_action( 'admin_enqueue_scripts', [ $this, 'cache_block_sources' ], PHP_INT_MAX );
+		}
+
+		add_action( 'activated_plugin', [ $this, 'clear_block_sources_cache' ] );
+		add_action( 'after_switch_theme', [ $this, 'clear_block_sources_cache' ] );
+		add_action( 'upgrader_process_complete', [ $this, 'clear_block_sources_cache' ] );
+	}
+
+	/**
+	 * Registers the google font style.
+	 *
+	 * @param array $args Array of arguments for registering a block type.
+	 * @return array Filtered block type args.
+	 */
+	public function capture_block_type_source( $args ) {
+		if ( isset( $this->get_block_sources()[ $args['name'] ] ) ) {
+			return $args;
+		}
+
+		$likely_culprit = $this->likely_culprit_detector->analyze_backtrace();
+
+		if ( in_array( $likely_culprit[ FileReflection::SOURCE_TYPE ], [ FileReflection::TYPE_PLUGIN, FileReflection::TYPE_MU_PLUGIN ], true ) ) {
+			$plugin = $this->plugin_registry->get_plugin_from_slug(
+				$likely_culprit[ FileReflection::SOURCE_NAME ],
+				FileReflection::TYPE_MU_PLUGIN === $likely_culprit[ FileReflection::SOURCE_TYPE ]
+			);
+
+			$likely_culprit['title'] = isset( $plugin['data']['Title'] ) ? $plugin['data']['Title'] : $likely_culprit[ FileReflection::SOURCE_NAME ];
+		} elseif ( FileReflection::TYPE_THEME === $likely_culprit[ FileReflection::SOURCE_TYPE ] ) {
+			$theme                   = wp_get_theme( $likely_culprit['name'] );
+			$likely_culprit['title'] = $theme->get( 'Name' ) ?: $likely_culprit[ FileReflection::SOURCE_NAME ];
+		} else {
+			$likely_culprit['title'] = __( 'WordPress core', 'amp' );
+		}
+
+		$this->block_sources[ $args['name'] ] = $likely_culprit;
+
+		return $args;
+	}
+
+	/**
+	 * Saves the block source data to cache.
+	 */
+	public function cache_block_sources() {
+		if ( wp_using_ext_object_cache() ) {
+			wp_cache_set( self::CACHE_KEY, $this->block_sources, __CLASS__, self::CACHE_TIMEOUT );
+		} else {
+			set_transient( __CLASS__ . self::CACHE_KEY, $this->block_sources, self::CACHE_TIMEOUT );
+		}
```

Comment:
When an external object cache is being used, transients are just wrappers for the object cache.

```suggestion
		set_transient( __CLASS__ . self::CACHE_KEY, $this->block_sources, self::CACHE_TIMEOUT );
```